fix: Skip Resolving Secrets From Restricted Namespace

Indresh2410 opened the pull request:
Fixes https://github.com/kedacore/keda/issues/4519

tomkerkhove reviewed (commented):
Thanks for the PR! Can you please update the release notes and add some tests?

JorTurFer reviewed (commented):
Shouldn't we check also the current namespace? I mean, if the ScaledObject/ScaledJob is in the allowed namespace, we should read the secret

zroubalik commented:
> Shouldn't we check also the current namespace? I mean, if the ScaledObject/ScaledJob is in the allowed namespace, we should read the secret

Well probably not "current" but the one that is restricted.

JorTurFer commented:
/run-e2e

tobotg commented on pkg/scaling/resolver/scale_resolvers.go:
```diff
@@ -286,7 +287,7 @@ func getTriggerAuthSpec(ctx context.Context, client client.Client, triggerAuthRe
 func resolveEnv(ctx context.Context, client client.Client, logger logr.Logger, container *corev1.Container, namespace string, secretsLister corev1listers.SecretLister) (map[string]string, error) {
 	resolved := make(map[string]string)
-
+	secretAccessRestricted, _ := strconv.ParseBool(util.GetRestrictSecretAccess())
```
Thanks @Indresh2410 
I think it's safer to use `isSecretAccessRestricted(logger)` instead of parsing the prop here.

tobotg commented on pkg/scaling/resolver/scale_resolvers.go:
```diff
@@ -302,7 +303,7 @@ func resolveEnv(ctx context.Context, client client.Client, logger logr.Logger, c
 					return nil, fmt.Errorf("error reading config ref %s on namespace %s/: %w", source.ConfigMapRef, namespace, err)
 				}
-			} else if source.SecretRef != nil && !isSecretAccessRestricted(logger) {
+			} else if source.SecretRef != nil && readSecrets(secretAccessRestricted, namespace) {
```
You may want to evaluate `readSecrets(secretAccessRestricted, namespace)` once and assign it to a boolean variable outside of the `for` loop

tobotg commented on pkg/scaling/resolver/scale_resolvers.go:
```diff
@@ -331,7 +332,7 @@ func resolveEnv(ctx context.Context, client client.Client, logger logr.Logger, c
 				// env is an EnvVarSource, that can be on of the 4 below
 				switch {
-				case envVar.ValueFrom.SecretKeyRef != nil && !isSecretAccessRestricted(logger):
+				case envVar.ValueFrom.SecretKeyRef != nil && readSecrets(secretAccessRestricted, namespace):
```
Same comment as above.

tobotg commented on pkg/scaling/resolver/scale_resolvers_test.go:
```diff
@@ -541,3 +541,54 @@ func TestResolveDependentEnv(t *testing.T) {
+				}},
+			},
+			expected: map[string]string{},
+		},
```
Is it possible to test "restricted secret access and namespace is keda" case?

Indresh2410 replied:
Hi @tobotg . When namespace is keda, we are trying to fetch secrets from secretsLIster. Any way we can mock secretLIster? I tried in local and I ended up getting below error after setting required env variables

panic: runtime error: invalid memory address or nil pointer dereference [recovered]

Any suggestions on this?

zroubalik replied:
Here you can see how we are mocking other stuff https://github.com/kedacore/keda/blob/bc921248e83d8c169d9ed9883e6c666979a9bcb2/Makefile#L174 - this is particular mocks scale interface of `k8s.io/clinet-go`. 

So the approach can be identical.

Indresh2410 replied:
Thanks @zroubalik for your suggestion. Have added test case for the same

tobotg reviewed (changes requested):
Thanks @Indresh2410 for this PR. Just a few comments.

JorTurFer commented:
> If KEDA_RESTRICT_SECRET_ACCESS is enabled, we can remove secrets from clusterrole [keda.sh/docs/2.10/operate/cluster/#restrict-secret-access](https://keda.sh/docs/2.10/operate/cluster/#restrict-secret-access)
> 
> But we would still require list/watch access to secrets right? [`main`/cmd/operator/main.go#L192-L193](https://github.com/kedacore/keda/blob/main/cmd/operator/main.go?rgh-link-date=2023-06-20T16%3A49%3A08Z#L192-L193)

yeah, this is because as default, KEDA can read secrets from all the cluster, restricting the secret access we don't block the access to the secrets, we just limit it to keda's namespace. In fact, list/watch isn't enough because the certificate management requires more verbs, but that's deployed in other role/rolebinding: https://github.com/kedacore/keda/blob/main/config/rbac/role.yaml#L122-L139

zroubalik reviewed (commented):
@Indresh2410 any update on this? we are about to release a new version.

Indresh2410 commented:
Hi. Sorry for delay. Will push changes by EOD

On Wed, Jun 21, 2023, 12:54 PM Zbynek Roubalik ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> @Indresh2410 <https://github.com/Indresh2410> any update on this? we are
> about to release a new version.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kedacore/keda/pull/4645#pullrequestreview-1489947561>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A7A3MRZWCHK4MAPWWEDK6ZLXMKORPANCNFSM6AAAAAAY2NKJCA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

zroubalik commented:
> Hi. Sorry for delay. Will push changes by EOD
> […](#)
> On Wed, Jun 21, 2023, 12:54 PM Zbynek Roubalik ***@***.***> wrote: ***@***.**** commented on this pull request. @Indresh2410 <https://github.com/Indresh2410> any update on this? we are about to release a new version. — Reply to this email directly, view it on GitHub <[#4645 (review)](https://github.com/kedacore/keda/pull/4645#pullrequestreview-1489947561)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/A7A3MRZWCHK4MAPWWEDK6ZLXMKORPANCNFSM6AAAAAAY2NKJCA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Thanks!

Indresh2410 commented:
> > If KEDA_RESTRICT_SECRET_ACCESS is enabled, we can remove secrets from clusterrole [keda.sh/docs/2.10/operate/cluster/#restrict-secret-access](https://keda.sh/docs/2.10/operate/cluster/#restrict-secret-access)
> > But we would still require list/watch access to secrets right? [`main`/cmd/operator/main.go#L192-L193](https://github.com/kedacore/keda/blob/main/cmd/operator/main.go?rgh-link-date=2023-06-20T16%3A49%3A08Z#L192-L193)
> 
> yeah, this is because as default, KEDA can read secrets from all the cluster, restricting the secret access we don't block the access to the secrets, we just limit it to keda's namespace. In fact, list/watch isn't enough because the certificate management requires more verbs, but that's deployed in other role/rolebinding: https://github.com/kedacore/keda/blob/main/config/rbac/role.yaml#L122-L139
> 
> KEDA will always have access to the secrets inside its own namespace

Thanks @JorTurFer .This helps 😄

tobotg commented on pkg/scaling/resolver/scale_resolvers_test.go:
```diff
@@ -592,3 +594,66 @@ func TestEnvWithRestrictSecretAccess(t *testing.T) {
+		},
+		Data: map[string][]byte{secretKey: []byte(secretData)},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			mock_secretNamespaceLister := mock_v1.NewMockSecretNamespaceLister(ctrl)
+			mock_secretNamespaceLister.EXPECT().Get(secretName).Return(secret, nil).AnyTimes()
+			mock_secretLister := mock_v1.NewMockSecretLister(ctrl)
+			mock_secretLister.EXPECT().Secrets(clusterNamespace).Return(mock_secretNamespaceLister).AnyTimes()
+			os.Setenv("KEDA_CLUSTER_OBJECT_NAMESPACE", "keda")
+			os.Setenv("KEDA_RESTRICT_SECRET_ACCESS", "true")
+			ctx := context.Background()
+			envMap, _ := resolveEnv(ctx, fake.NewClientBuilder().Build(), logf.Log.WithName("test"), test.container, clusterNamespace, mock_secretLister)
+			if diff := cmp.Diff(envMap, test.expected); diff != "" {
+				t.Errorf("Returned env map is different: %s", diff)
+			}
+		})
+	}
```
Since the mock has no constraint or specific return value, what about having it out of the loop?

```suggestion
        ctrl := gomock.NewController(t)
	mock_secretNamespaceLister := mock_v1.NewMockSecretNamespaceLister(ctrl)
	mock_secretNamespaceLister.EXPECT().Get(secretName).Return(secret, nil).AnyTimes()
	mock_secretLister := mock_v1.NewMockSecretLister(ctrl)
	mock_secretLister.EXPECT().Secrets(clusterNamespace).Return(mock_secretNamespaceLister).AnyTimes()
	for _, test := range tests {
		test := test
		t.Run(test.name, func(t *testing.T) {
			os.Setenv("KEDA_CLUSTER_OBJECT_NAMESPACE", "keda")
			os.Setenv("KEDA_RESTRICT_SECRET_ACCESS", "true")
			ctx := context.Background()
			envMap, _ := resolveEnv(ctx, fake.NewClientBuilder().Build(), logf.Log.WithName("test"), test.container, clusterNamespace, mock_secretLister)
			if diff := cmp.Diff(envMap, test.expected); diff != "" {
				t.Errorf("Returned env map is different: %s", diff)
			}
		})
	}
```

Indresh-Prakash replied:
Thanks @tobotg for suggestion. Have addressed the same 😃

Indresh2410 replied:
@tobotg thanks for your suggestion. Have addressed the same 😃

tobotg replied:
Thanks @Indresh2410 👍

zroubalik commented:
@Indresh2410 could you please rebase this PR and fix the conflict in Changelog?

zroubalik reviewed (commented):
@Indresh2410 could you please fix the path here? `pkg/mock/mock_secretLister/mock_interfaces.go`

make it lowercase only: `pkg/mock/mock_secretlister/mock_interfaces.go`

tobotg commented on Makefile:
```diff
@@ -171,6 +171,9 @@ pkg/mock/mock_scaling/mock_executor/mock_interface.go: pkg/scaling/executor/scal
 	$(MOCKGEN) -destination=$@ -package=mock_executor -source=$^
 pkg/mock/mock_scaler/mock_scaler.go: pkg/scalers/scaler.go
 	$(MOCKGEN) -destination=$@ -package=mock_scalers -source=$^
+pkg/mock/mock_secretLister/mock_interfaces.go: vendor/k8s.io/client-go/listers/core/v1/secret.go
```
@Indresh2410 I guess that `pkg/mock/mock_secretlister/mock_interfaces.go` needs to be added to `mockgen-gen` target above. @zroubalik could you please confirm? Thanks

zroubalik replied:
@tobotg good catch, you are right!

zroubalik commented:
@Indresh2410 do you think you'll be able to tackle the rest of the problems today? We plan to release a new version in a few hours.

zroubalik commented:
> Yes. Will address changes in next 1 hour. Hope it's fine for you
> […](#)
> On Thu, Jun 22, 2023, 7:06 PM Zbynek Roubalik ***@***.***> wrote: @Indresh2410 <https://github.com/Indresh2410> do you think you'll be able to tackle the rest of the problems today? We plan to release a new version in a few hours. — Reply to this email directly, view it on GitHub <[#4645 (comment)](https://github.com/kedacore/keda/pull/4645#issuecomment-1602653364)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/A7A3MR7XJAKUA7PDIP6KRR3XMRC6FANCNFSM6AAAAAAY2NKJCA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Thanks!

Indresh2410 commented:
> > Yes. Will address changes in next 1 hour. Hope it's fine for you
> > […](#)
> > On Thu, Jun 22, 2023, 7:06 PM Zbynek Roubalik _**@**_._**> wrote: @Indresh2410 https://github.com/Indresh2410 do you think you'll be able to tackle the rest of the problems today? We plan to release a new version in a few hours. — Reply to this email directly, view it on GitHub <[#4645 (comment)](https://github.com/kedacore/keda/pull/4645#issuecomment-1602653364)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/A7A3MR7XJAKUA7PDIP6KRR3XMRC6FANCNFSM6AAAAAAY2NKJCA . You are receiving this because you were mentioned.Message ID: _**@**_.**_>
> 
> Thanks!

HI @zroubalik . I've addressed all review comments. Please let me know if any other change is required

zroubalik reviewed (commented):
@Indresh2410 now there are both `pkg/mock/mock_secretLister/mock_interfaces.go` and `pkg/mock/mock_secretlister/mock_interfaces.go`.

Could you please fix that?

Indresh2410 commented:
> @Indresh2410 now there are both `pkg/mock/mock_secretLister/mock_interfaces.go` and `pkg/mock/mock_secretlister/mock_interfaces.go`.
> 
> Could you please fix that?

Have fixed it @zroubalik . Thanks

zroubalik commented:
/run-e2e internal

zroubalik reviewed (approved):
LGTM

zroubalik reviewed (commented):
@Indresh2410 unfortunately there are some static errors and also errors in the unit test. Could you please check?

zroubalik commented on CHANGELOG.md:
```diff
@@ -88,6 +88,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **Grafana Dashboard**: Fix HPA metrics panel to use range instead of instant ([#4513](https://github.com/kedacore/keda/pull/4513))
 - **Grafana Dashboard**: Fix HPA metrics panel by replacing $namepsace to $exported_namespace due to label conflict ([#4539](https://github.com/kedacore/keda/pull/4539))
 - **Kafka Scaler**: Add back `strings.TrimSpace()` function for saslAuthType ([#4689](https://github.com/kedacore/keda/issues/4689))
+- **General**: Skip resolving secrets if namespace is restricted ([#4634](https://github.com/kedacore/keda/pull/4645))
```
```suggestion
- **General**: Skip resolving secrets if namespace is restricted ([#4519](https://github.com/kedacore/keda/issues/4519))
```

Indresh2410 commented:
Hi @zroubalik . Looks like static check is still failing and I would like to take some time to fix the issues. Hope we can merge it to next release ?

Sorry for inconvenience caused

zroubalik commented:
@Indresh2410  are you able to run the test locally?
```
=== RUN   TestEnvWithRestrictSecretAccess
=== RUN   TestEnvWithRestrictSecretAccess/env_reference_secret_key
    scale_resolvers_test.go:592: Returned env map is different:   map[string]string(
        - 	nil,
        + 	{},
          )
=== RUN   TestEnvWithRestrictSecretAccess/env_reference_secret_name
    scale_resolvers_test.go:592: Returned env map is different:   map[string]string(
        - 	nil,
        + 	{},
          )
--- FAIL: TestEnvWithRestrictSecretAccess (0.00s)
    --- FAIL: TestEnvWithRestrictSecretAccess/env_reference_secret_key (0.00s)
    --- FAIL: TestEnvWithRestrictSecretAccess/env_reference_secret_name (0.00s)
=== RUN   TestEnvWithRestrictedNamespace
=== RUN   TestEnvWithRestrictedNamespace/env_reference_secret_key
    scale_resolvers_test.go:655: Returned env map is different:   map[string]string(
        - 	nil,
        + 	{"test-env-key": "secretDataHere"},
          )
=== RUN   TestEnvWithRestrictedNamespace/env_reference_secret_name
    scale_resolvers_test.go:655: Returned env map is different:   map[string]string(
        - 	nil,
        + 	{"mysecretkey": "secretDataHere"},
          )
--- FAIL: TestEnvWithRestrictedNamespace (0.00s)
    --- FAIL: TestEnvWithRestrictedNamespace/env_reference_secret_key (0.00s)
    --- FAIL: TestEnvWithRestrictedNamespace/env_reference_secret_name (0.00s)
FAIL
	github.com/kedacore/keda/v2/pkg/scaling/resolver	coverage: 35.2% of statements
FAIL	github.com/kedacore/keda/v2/pkg/scaling/resolver	0.104s
```

zroubalik commented:
> > @Indresh2410 are you able to run the test locally?
> > ```
> > === RUN   TestEnvWithRestrictSecretAccess
> > === RUN   TestEnvWithRestrictSecretAccess/env_reference_secret_key
> >     scale_resolvers_test.go:592: Returned env map is different:   map[string]string(
> >         - 	nil,
> >         + 	{},
> >           )
> > === RUN   TestEnvWithRestrictSecretAccess/env_reference_secret_name
> >     scale_resolvers_test.go:592: Returned env map is different:   map[string]string(
> >         - 	nil,
> >         + 	{},
> >           )
> > --- FAIL: TestEnvWithRestrictSecretAccess (0.00s)
> >     --- FAIL: TestEnvWithRestrictSecretAccess/env_reference_secret_key (0.00s)
> >     --- FAIL: TestEnvWithRestrictSecretAccess/env_reference_secret_name (0.00s)
> > === RUN   TestEnvWithRestrictedNamespace
> > === RUN   TestEnvWithRestrictedNamespace/env_reference_secret_key
> >     scale_resolvers_test.go:655: Returned env map is different:   map[string]string(
> >         - 	nil,
> >         + 	{"test-env-key": "secretDataHere"},
> >           )
> > === RUN   TestEnvWithRestrictedNamespace/env_reference_secret_name
> >     scale_resolvers_test.go:655: Returned env map is different:   map[string]string(
> >         - 	nil,
> >         + 	{"mysecretkey": "secretDataHere"},
> >           )
> > --- FAIL: TestEnvWithRestrictedNamespace (0.00s)
> >     --- FAIL: TestEnvWithRestrictedNamespace/env_reference_secret_key (0.00s)
> >     --- FAIL: TestEnvWithRestrictedNamespace/env_reference_secret_name (0.00s)
> > FAIL
> > 	github.com/kedacore/keda/v2/pkg/scaling/resolver	coverage: 35.2% of statements
> > FAIL	github.com/kedacore/keda/v2/pkg/scaling/resolver	0.104s
> > ```
> 
> Exporting variables and running tests worked fine for me @zroubalik

@Indresh2410 your branch is faling for me locally as well. It is fine if you don't have a time to fix it now, we can postpone it to the next release. Just let me know.

Indresh2410 commented:
> > > @Indresh2410 are you able to run the test locally?
> > > ```
> > > === RUN   TestEnvWithRestrictSecretAccess
> > > === RUN   TestEnvWithRestrictSecretAccess/env_reference_secret_key
> > >     scale_resolvers_test.go:592: Returned env map is different:   map[string]string(
> > >         - 	nil,
> > >         + 	{},
> > >           )
> > > === RUN   TestEnvWithRestrictSecretAccess/env_reference_secret_name
> > >     scale_resolvers_test.go:592: Returned env map is different:   map[string]string(
> > >         - 	nil,
> > >         + 	{},
> > >           )
> > > --- FAIL: TestEnvWithRestrictSecretAccess (0.00s)
> > >     --- FAIL: TestEnvWithRestrictSecretAccess/env_reference_secret_key (0.00s)
> > >     --- FAIL: TestEnvWithRestrictSecretAccess/env_reference_secret_name (0.00s)
> > > === RUN   TestEnvWithRestrictedNamespace
> > > === RUN   TestEnvWithRestrictedNamespace/env_reference_secret_key
> > >     scale_resolvers_test.go:655: Returned env map is different:   map[string]string(
> > >         - 	nil,
> > >         + 	{"test-env-key": "secretDataHere"},
> > >           )
> > > === RUN   TestEnvWithRestrictedNamespace/env_reference_secret_name
> > >     scale_resolvers_test.go:655: Returned env map is different:   map[string]string(
> > >         - 	nil,
> > >         + 	{"mysecretkey": "secretDataHere"},
> > >           )
> > > --- FAIL: TestEnvWithRestrictedNamespace (0.00s)
> > >     --- FAIL: TestEnvWithRestrictedNamespace/env_reference_secret_key (0.00s)
> > >     --- FAIL: TestEnvWithRestrictedNamespace/env_reference_secret_name (0.00s)
> > > FAIL
> > > 	github.com/kedacore/keda/v2/pkg/scaling/resolver	coverage: 35.2% of statements
> > > FAIL	github.com/kedacore/keda/v2/pkg/scaling/resolver	0.104s
> > > ```
> > 
> > 
> > Exporting variables and running tests worked fine for me @zroubalik
> 
> @Indresh2410 your branch is faling (the test) for me locally as well. It is fine if you don't have a time to fix it now, we can postpone it to the next release. Just let me know.

Yes @zroubalik . We'll push it as part of next release. Will sort out issues sometime in next few days. Thanks for quick reviews 😃

Indresh2410 commented:
Hi @zroubalik . Sry for inconvenience caused. Have raised another PR with same changes 
https://github.com/kedacore/keda/pull/4735

Closing this in favour of the above
